### ISSUE TITLE
train with DDP is slower

### ISSUE THREAD

**Issue by wuzhi19931128** [PhilJd/contiguous_pytorch_params#2]
When I just add params.py to my code, it run slower inTitan _XP. I found that when i define optimizer after DDP and it will keep the time almost same, but when I followed the readme.py and define optimizer before DDP, it run slower. What's wrong in my code? Thanks for your answer.
[pp_imagenet.txt](https://github.com/PhilJd/contiguous_pytorch_params/files/5231601/pp_imagenet.txt)

**Comment by PhilJd** [comment]
Hi,
I usually found it necessary to move the model to a cuda device first before creating the contiguous parameters and DDP. Could you try that? 
Does the loss go down in both cases?

**Comment by wuzhi19931128** [comment]
I use the commond "python -m torch.distributed.launch --nproc_per_node 8 pp_imagenet --save_dit . --cont True" to run it. The    args "cont" indicate whether to use contiguous_pytorch_params.

**Comment by wuzhi19931128** [comment]
Sorry for that mistaken， I will re-check my code .

**Comment by wuzhi19931128** [comment]
ok，I found that define optimizer after DDP  or before DDP can both reduce the time of optimizer.step(). This is an interesting little trick.

**Comment by PhilJd** [comment]
Thanks for reporting back! :)
Hm, interesting to hear that this also works with initializing after DDP. I'm curious, did you verify that this trains the model on all GPUs, i.e., that gradients are still synced across GPUs? (Maybe by setting the loss to zero for the main GPU and checking that the gradients are non-zero for the other ops)

**Comment by wuzhi19931128** [comment]
I'm sorry to see your reply so late. This is a problem completely unrelated to the optimizer. 
when you use DDP ，Pytorch  will synchronize gradients automatically, whether it is 0 or not. In fact, in most cases, the gradient of different cards is different, Pytorch will average them. 
Maybe your question is whether you can change the weight of one gpu to be different from that of other gpus. yes, the weights in different gpus can be different. Pytorch synchronize gradients and updated weights in each GPU. when weights is differents, it still works. You can find different model in different gpus but they don't konw they are different.

**Comment by PhilJd** [comment]
I meant something slightly different ;)
DDP creates a synchronization op between the parameters that are present when passing the model to the DDP constructor. When constructing the contiguous optimizer after DDP, this replaces the parameter buffers, potentially destroying the syncing connection to the other GPUs (That's the part I'm unsure about and wanted to verify). This means, all of your cards would compute gradients but not sync them anymore. The model might still train though as each card still gets its own gradients, but training would be slower as the effective batch size is smaller. What I suggested is to manually set the gradients of the main GPU to zero (e.g. by using its own prediction as ground truth in the loss). If syncing still works, then the gradients on the main GPU should be non-zero as it gets non-zero gradients from the other GPUs. If the sync connection was broken by constructing the contiguous parameters after DDP, the gradients on the main GPU should remain zero.
